Preparing convict 6 (#353)
mozilla/node-convict: packages/convict/README.md

config.loadFile(['./prod.json', './config.json']);
```

> Each setting in the schema has the following possible properties, each aiding in convict's goal of being more robust and collaborator friendly.

Reviewer: ```suggestion
Each setting in the schema has the following possible properties, each aiding in the convict's goal of being more robust and collaborator friendly.
```

### Convict properties

>  - **Type information**: the `format` property specifies either a built-in convict format (`ipaddress`, `port`, `int`, etc.), or it can be a function to check a custom format. During validation, if a format check fails it will be added to the error report.

Reviewer: ```suggestion
 - **Type information**: the `format` property specifies either a built-in convict format (`ipaddress`, `port`, `int`, etc.) or it can be a function to check a custom format. During validation, if a format check fails it will be added to the error report.
```

>  - **Default values**: Is a default value before this value will be rewrite by another getter. 

Reviewer: ```suggestion
 - **Default values**: Is a default value before this value will be overwritten by another getter. 
```

### Coercion

Convict will automatically coerce environmental variables from strings to their proper types when importing them.
For instance, values with the format `int`, `nat`, `port`, or `Number` will become numbers after a straight forward
`parseInt` or `parseFloat`.

### Precedence order

> When merging configuration values from different sources, Convict follows precedence rules depending of the getters order.

Reviewer: ```suggestion
When merging configuration values from different sources, Convict follows precedence rules depending on the getters' order.
```

The default getters order, from lowest to highest:

```javascript
convict.getGettersOrder();
// ['default', 'value', 'env', 'arg', 'force']
```

1. Use `default` property set in schema
2. **Value** used with:
    - `config.loadFile(file)` and `config.load(json)` ;
    - `config.set(name, value, false, true)`.
3. Use the environment variable (only used when `env` property is set in schema)
4. Use the commandline arguments (only used when `arg` property is set in schema)
5. **Force** used with:
    - With: `config.set(name, value, true)` (permanent) ;
    - With: `config.set(name, value)` (can be undo with `config.refreshGetters()`).

This order means that if schema defines parameter to be taken from an environment variable and environment variable is set then you cannot override it with `config.loadFile(file)` or `config.load(json)`.

```javascript
process.env.PORT = 8080; // environment variable is set
const config = convict({
  port: {
    default: 3000,
    env: 'PORT'
  }
});
config.load({ port: 9000 });
console.log(config.get('port')); // still 8080 from env
```

### Overriding Environment variables and Command line arguments

> Convict allows to override Environment variables and Command line arguments.

Reviewer: ```suggestion
Convict allows to override Environment variables and Command-line arguments.
```

convict.addParser({extension: 'json', parse: require('json5').parse});
```

## API convict (global)

> Some function are only global, like : addParser, addFormat, addGetter...

Reviewer: ```suggestion
Some functions are only global, like `addParser`, `addFormat`, `addGetter`, etc.
```

### convict.addParser(parser)

>  - **parser**: is an Object or an Array contains containing Object.

Reviewer: `contains containing Object` - not sure what did you mean here

Author: - **parser**: is an Object or an Array containing Objects. maybe better

Adds new parsers for custom file extensions. Parser should be an Object.
E.g.:
```javascript
// Allow comments in JSON file (with JSON5)
convict.addParser({ extension: 'json5', parse: require('json5').parse });
```

### convict.addFormat(format) or convict.addFormat(name, validate, coerce[, rewrite = false])

> Adds a new custom format, `format` being an object, see example below. `rewrite = true`

Reviewer: ```suggestion
Adds a new custom format, `format` is an object, see example below. `rewrite = true`
```

will try to call each getter function to get a value (different of `undefined`), then `stopPropagation()`
stops the getter calling loop.
E.g.:
```javascript
convict.addGetter({
  name: 'file',
  getter: (value, schema, stopPropagation) => fs.readFileSync(value, 'utf-8').toString(),
  usedOnlyOnce: true // use file only once
});

convict.addGetter({
  property: 'accept-undefined',
  getter: (value, schema, stopPropagation) => {
    stopPropagation();
    return undefined;
  }
});
```

### convict.addGetters(getters)

Adds new custom getter, `getter` being an object whose keys are the new custom
getter names.
E.g.:
```javascript
convict.addGetters([
  /* example to rewrite 'env' getter: */
  { name: 'env', getter: (val) => schema._cvtCoerce(this.getEnv()[val]), usedOnlyOnce: false, rewrite: true },
  { name: 'getter2', getter: (val) => val }
]);
```

### convict.getGettersOrder()

Returns array containing getter names sorted by priority (ascending order).
E.g.:
```javascript
convict.getGettersOrder();
// ['default', 'value', 'env', 'arg', 'force']
```

Also see: [`config.getGettersOrder()`](#TEMP_LINK)

### convict.sortGetters(newOrder)

> Sort getter depending of array order, priority uses ascending order. You have

Reviewer: ```suggestion
Sort getter depending on array order, priority uses ascending order. You have
```

    'doc': 'Log file path',
    'format': String,
    'default': '/tmp/app.log'
  }
});

// or
config = convict('/some/path/to/a/config-schema.json');
```

### config.get(name)

Returns the current value of the `name` property. `name` can use dot notation to reference nested values.
E.g.:
```javascript
config.get('db.host');
// or
config.get('db').host;
// also
config.get('db[0]');
// with dot:
config.get('db["www.airbus.com"]'); { 'db': { 'www.airbus.com': 'air company'} }
// in the first level
config.get("['foo.bar']"); // { 'foo.bar': 'baz' }
```

### config.getOrigin(name)

Returns the current getter name of the `name` value origin. `name` can use dot notation to reference nested values.
E.g.:
```javascript
config.getOrigin('db.host');
```

### config.getGettersOrder()

Local (configuration instance) version of : [`convict.getGettersOrder()`](#TEMP_LINK)

### config.sortGetters(newOrder)

Local (configuration instance) version of : [`convict.sortGetters()`](#TEMP_LINK)

Sort getter depending of array order, priority uses ascending order.

### config.refreshGetters()

Reclone global getters config to local getters config and update configuration object value depending

> of new getters order.

Reviewer: ```suggestion
on new getters' order.
```

convict.getGettersOrder(); // ['value', 'default', 'arg', 'env', 'force']

conf.refreshGetters(); // refresh and apply global change to local

conf.getGettersOrder(); // ['value', 'default', 'arg', 'env', 'force']
```

### config.default(name)

Returns the default value of the `name` property. `name` can use dot notation to reference nested values.
E.g.:
```javascript
config.default('server.port');
```

### config.reset(name)

Resets a property to its default value as defined in the schema.
E.g.:
```javascript
config.reset('server.port');
```

### config.has(name)

Returns `true` if the property `name` is defined, or `false` otherwise.
E.g.:
```javascript
if (config.has('some.property')) {
  // Do something
}
```

### config.set(name, value[, priority = false, respectPriority = false])

Sets the value of `name` to value. `name` can use dot notation to reference
nested values, e.g. `"db.port"`. If objects in the chain don't yet exist,
they will be initialized to empty objects.

**priority**: Optional, can be a boolean or getter name (a string). You must
declare this property in the schema to use this option. `set` will change

> the property getter origin depending of `priority` value:

Reviewer: ```suggestion
the property getter origin depending on `priority` value:
```

Validates `config` against the schema used to initialize it. All errors are
collected and thrown or displayed at once.

#### options :

1. `allowed`: Any properties specified in config files that are not declared in

>    the schema will print a warning or throw an error depending of this setting:

Reviewer: ```suggestion
   the schema will print a warning or throw an error depending on this setting:
```

   - `warn`: is the default behavior, will print a warning.
   - `strict`: will throw errors. This is to ensure that the schema and the config
     files are in sync.

2. `output`: You can replace the default output `console.log`
   by your own output function. You can use [debug module][debug] like this:

> E.g.:

Reviewer: ```suggestion
```

The map of environment variables. Defaults to process.env unless an override is specified using the env key of the second argument (options) argument of the convict function.

## API schema property

> Schema property is accessible by second argument of the callback of getter `getter(value, schema, stopPropagation)` and format `validate(value, schema)`.

Reviewer: ```suggestion
Schema property is accessible by the second argument of the callback of getter `getter(value, schema, stopPropagation)` and format `validate(value, schema)`.
```

> ### schema[`doc` || `defaut` || `format` || ...]

Reviewer: ```suggestion
### schema[`doc` || `default` || `format` || ...]
```

> You will find the properties that you defined in your schema (like `doc`, `defaut`, `format`, etc...).

Reviewer: ```suggestion
You will find the properties that you defined in your schema (like `doc`, `default`, `format`, etc...).
```
